rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,491 @@
+possible to avoid termination of the connection.
+
+The SDK might also send PING frames to the server and expect PONG frame. If the
+server does not respond to SDK-initiated PING request, the SDK should close
+connection, and reconnect after `app_telemetry_backoff` interval.
+
+## Wire Protocol
+
+The structure of the protocol is intentionally simple, and yet allows for
+future extension.
+
+The SDK leverages WebSocket framing, so does not need to have special fields
+for the payload size.
+
+First byte is reserved for Command codes in requests from the server, and for
+Status for SDK responses.
+
+This RFC defines the following types of the commands.
+
+| Name            | Value  | Description                                                                  |
+|-----------------|--------|------------------------------------------------------------------------------|
+| `GET_TELEMETRY` | `0x00` | Retrieve Application Telemetry from SDK in Prometheus Text Exposition Format |
+
+And the following statuses:
+
+| Name              | Value  | Description                                                |
+|-------------------|--------|------------------------------------------------------------|
+| `SUCCESS`         | `0x00` | No errors has been occurred                                 |
+| `UNKNOWN_COMMAND` | `0x01` | The SDK does not support operation requested by the server |
+
+
+### `GET_TELEMETRY`
+
+The request must not have payload.
+
+The response might have payload if the SDK has metrics to report.
+
+The SDK must reset local counters after successful scheduling response for the
+network delivery.
+
+The format of the response follows the [Text-based exposition format of the
+Prometheus](https://prometheus.io/docs/instrumenting/exposition_formats/#text-based-format),
+and includes metrics and histograms collected by the SDK since the previous
+report.
+
+#### Metrics
+
+    sdk_kv_r_utimedout{agent="couchbase-net-sdk/2.4.5.0 (clr/4.0.30319.42000) (os/Microsoft Windows NT 10.0.16299.0)", id="66388CF5BFCF7522/18CC8791579B567C", node="node1", bucket="travel-sample"} 1 1695747260
+
+Where
+
+* `sdk_kv_r_utimedout` is a name of the metric, one of the:
+
+  | Metric | Description |
+  | ---- | ---- |
+  | `sdk_{service}_r_utimedout` | Unambiguous timeout |
+  | `sdk_{service}_r_atimedout` | Ambiguous timeout |
+  | `sdk_{service}_r_canceled` | Canceled |
+  | `sdk_{service}_r_total` | Total number of operations |
+
+* `{agent="couchbase-net-sdk/2.4.5.0 (clr/4.0.30319.42000) (os/Microsoft Windows NT 10.0.16299.0)", id="66388CF5BFCF7522/18CC8791579B567C", node="node1", bucket="travel-sample"}` is a list of the tags.
+
+  | Label | Description |
+  | ---- | ---- |
+  | `agent` | agent string of the SDK, as it is used everywhere (HELLO message, logs etc.) |
+  | `id` | SDK identifier string as specified by [SDK-RFC #35](0035-rto.md#client--connection-ids) in section "Client / Connection IDs" |
```

Comment:
My understanding is that this is KV-only. If that's the case I think it should be made clear in the RFC that this field is only expected to be present for KV requests.

Reply:
HTTP services also have connections IDs

Reply:
Apologies, I had RFC-67 in mind which uses Connection IDs only for KV. RFC-35 does indeed specify them for all services.

However, it could be problematic to implement in Go, as we don't manage the HTTP connection pool ourselves. We rely on the standard library's HTTP client.

In addition, I think that recording a Connection ID for HTTP services could result in significant combinatorial explosion, even more so compared to KV, as the number of HTTP connections can be unbounded, whereas for KV we have a fixed number of connections.

Reply:
Java SDK lets users configure a very large number of KV connections -- so the combinatorial explosion concern is important for KV too.